packages/core/src/lib/TwaGenerator.ts

```diff
@@ -296,10 +342,38 @@ export class TwaGenerator {
     }
     progress.update();
 
-    await Promise.all(twaManifest.shortcuts.map((shortcut: ShortcutInfo, i: number) => {
-      const imageDirs = SHORTCUT_IMAGES.map(
-          (imageDir) => ({...imageDir, dest: `${imageDir.dest}shortcut_${i}.png`}));
-      return this.generateIcons(shortcut.chosenIconUrl, targetDirectory, imageDirs);
+    await Promise.all(twaManifest.shortcuts.map(async (shortcut: ShortcutInfo, i: number) => {
+      const assetName = shortcut.assetName(i);
+      const monochromeAssetName = `${assetName}_monochrome`;
+      const maskableAssetName = `${assetName}_maskable`;
+      const templateArgs = {assetName, monochromeAssetName, maskableAssetName};
+
+      if (shortcut.chosenMonochromeIconUrl) {
+        await this.applyTemplateMap(
+            templateDirectory, targetDirectory,
+            shortcutMonochromeTemplateFileMap(assetName), templateArgs);
+
+        const monochromeImages = shortcutImages(monochromeAssetName);
+
+        const baseMonochromeIcon = await this.fetchIcon(shortcut.chosenMonochromeIconUrl);
+        const monochromeIcon = await this.monochromeFilter(baseMonochromeIcon, twaManifest);
+
+        return await Promise.all(monochromeImages.map((iconDef) => {
+          return this.generateIcon(monochromeIcon, targetDirectory, iconDef);
+        }));
+      }
+
+      if (shortcut.chosenMaskableIconUrl) {
+        await this.applyTemplateMap(
+            templateDirectory, targetDirectory,
+            shortcutMaskableTemplateFileMap(assetName), templateArgs);
+
+        const maskableImages = shortcutImages(maskableAssetName);
+        await this.generateIcons(shortcut.chosenMaskableIconUrl, targetDirectory, maskableImages);
+      }
+
+      const images = shortcutImages(assetName);
+      return this.generateIcons(shortcut.chosenIconUrl!, targetDirectory, images);
```

Comment:
It looks to me that `shortcut.chosenIconUrl` could be undefined here, as it's not required on the `ShortcutInfo` constructor anymore. Maybe I'm missing something, but I don't see a check that ensures `chosenIconUrl` is not null and this could lead to an error error in `generateIcons`?

Reply:
It should be defined here as monochrome icon is undefined on this branch. I'll add a runtime check too.

Reply:
Got it - I had missed the early return and thought all 3 icons could be generated. My understanding then is:

- We'll either have `chosenIconUrl` or `monoChromeIconUrl` defined. Having both is possible but doesn't really make sense as the first will be ignored when the second is defined.
- `monoChromeIconUrl` is ignored if `monoChromeIconUrl` is defined.

Is this correct?

Reply:
Yes, that's correct. I've tried to check this behaviour in the ShortcutInfo constructor too.